File: src/iOS/NeurofeedbackApp/Devices/Models/EEGModels.swift
//
//  EEGModels.swift
//  NeurofeedbackApp
//
//  Created by Shishir Mishra on 3/1/2024.
//

import Foundation

enum EEGFrequency : String, CaseIterable,Encodable {
    case all
    case theta
    case alpha
    case beta
    case gamma
    
    var id :String {
        rawValue
    }
}

//electrical activity or voltage measured by each electrode at that specific time
struct EEGChannel :RawRepresentable, Hashable,Encodable {
    let rawValue: String
    // in-built channel in Muse
    static let af7 = EEGChannel(rawValue: "AF7")
    static let af8 = EEGChannel(rawValue: "AF8")
    static let fpz = EEGChannel(rawValue: "FPZ")
    static let t9 = EEGChannel(rawValue: "T9")
    static let t10 = EEGChannel(rawValue: "T10")
    // Additional possibble channel for external sensrs
    
    static let cz = EEGChannel(rawValue: "CZ")
    static let fz = EEGChannel(rawValue: "FZ")
    static let f3 = EEGChannel(rawValue: "F3")
    static let f4 = EEGChannel(rawValue: "F4")
    static let o1 = EEGChannel(rawValue: "O1")
}

struct EEGSeries : Identifiable,Encodable {
    var id :Date {
        timestamp
    }
    let timestamp : Date
    
    private let readingsDictionary :[EEGChannel: EEGReading]
   
    init(timestamp: Date, readings: [EEGReading]) {
        self.timestamp = timestamp
        self.readingsDictionary = Dictionary(uniqueKeysWithValues: readings.map {($0.channel, $0)})
    }
    func getChannels() -> [EEGChannel] {
        return Array(readingsDictionary.keys)
    }
    
    func getReadingForChannel(channel:EEGChannel) -> EEGReading {
        guard let reading  = readingsDictionary[channel] else {
            preconditionFailure("No reading found for channel")
        }
        return reading
    }
}

struct EEGReading:Encodable {
    
    let channel: EEGChannel
    let value : Double
    
    init(channel: EEGChannel, value: Double) {
        self.channel = channel
        self.value = value
    }
}

struct EEGRecording :Encodable{
    let baseTime:TimeInterval
    let data:[EEGFrequency:[EEGReading]]
    let sessionId : UUID
    let userId : UUID
    
    init(baseTime: TimeInterval, data: [EEGFrequency : [EEGReading]], sessionId: UUID, userId: UUID) {
        self.baseTime = baseTime
        self.data = data
        self.sessionId = sessionId
        self.userId = userId
    }   
}

File: src/Server/DotNet.Commons/NeurofeedbackApp.Commons/Models/EEGModels.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace NeurofeedbackApp.Commons.Models
{
    public enum EEGFrequency
    {
        All,
        Theta,
        Alpha,
        Beta,
        Gamma
    }

    public record EEGReading(
     [property: JsonProperty("channel")] string Channel,
        [property: JsonProperty("value")] double Value
);

    public struct EEGChannel : IEquatable<EEGChannel>, IEquatable<string>
    {
        public string RawValue { get; }

        // In-built channel in Muse
        public static EEGChannel Af7 { get; } = new EEGChannel("AF7");
        public static EEGChannel Af8 { get; } = new EEGChannel("AF8");
        public static EEGChannel Fpz { get; } = new EEGChannel("FPZ");
        public static EEGChannel T9 { get; } = new EEGChannel("T9");
        public static EEGChannel T10 { get; } = new EEGChannel("T10");

        // Additional possible channels for external sensors
        public static EEGChannel Cz { get; } = new EEGChannel("CZ");
        public static EEGChannel Fz { get; } = new EEGChannel("FZ");
        public static EEGChannel F3 { get; } = new EEGChannel("F3");
        public static EEGChannel F4 { get; } = new EEGChannel("F4");
        public static EEGChannel O1 { get; } = new EEGChannel("O1");

        private EEGChannel(string rawValue)
        {
            RawValue = rawValue;
        }

        public bool Equals(EEGChannel other)
        {
            return RawValue == other.RawValue;
        }

        public bool Equals(string other)
        {
            return RawValue == other;
        }

        public override bool Equals(object obj)
        {
            return obj is EEGChannel channel && Equals(channel);
        }

        public override int GetHashCode()
        {
            return RawValue.GetHashCode();
        }

        public static bool operator ==(EEGChannel left, EEGChannel right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(EEGChannel left, EEGChannel right)
        {
            return !(left == right);
        }
    }

    // Could not use record to string to EEGFrequency 
    // come back to it later
    public class EEGFrequencyReadings
    {
        [JsonProperty("readings")]
        public List<EEGReading> Readings { get; set; }

        [JsonProperty("frequency")]
        [JsonConverter(typeof(StringEnumConverter))]
        public EEGFrequency EEGFrequency { get; set; }
    }

    public record EEGRecording(
        [property: JsonProperty("sessionId")] string SessionId,
        [property: JsonProperty("data")] IReadOnlyList<EEGFrequencyReadings> Data,
        [property: JsonProperty("userId")] string UserId,
        [property: JsonProperty("baseTime")] double BaseTime
);

}
